File: ios/Sources/TaigiKeyboard/Settings/OneHandedMode.swift
import Foundation

/// One-handed keyboard mode: the key area narrows to 80% and docks to one edge.
///
/// Mirrors Android `ime/core/settings/OneHandedMode.kt`; [rawValue] is the persisted
/// string, identical on both platforms.
enum OneHandedMode: String, CaseIterable {
    case off
    case left
    case right

    // CROSS-PLATFORM INVARIANT — mirrors android/app/src/main/java/com/siansiansu/taigikeyboard/ime/core/settings/OneHandedMode.kt
    // KEY_AREA_FRACTION; KeyboardKit docks the key rows at this share. Drift causes silent divergence.
    static let keyAreaFraction: CGFloat = 0.8

    /// The opposite side; `.off` stays `.off`.
    var flipped: OneHandedMode {
        switch self {
        case .off: .off
        case .left: .right
        case .right: .left
        }
    }
}

/// The action behind the toolbar keyboard button — the most recent pick from its
/// long-press callout (choosing Normal there leaves it unchanged).
///
/// Mirrors Android `KeyboardToolbarAction`; [rawValue] is the persisted string.
enum KeyboardToolbarAction: String, CaseIterable {
    case dismiss
    case left
    case right

    /// The one-handed mode this action docks to; `nil` for `.dismiss`.
    var mode: OneHandedMode? {
        switch self {
        case .dismiss: nil
        case .left: .left
        case .right: .right
        }
    }

    /// Mode after a tap on the toolbar button while in [current]; `nil` = dismiss the keyboard.
    /// A side action toggles between that side and `.off`.
    func tapResult(current: OneHandedMode) -> OneHandedMode? {
        guard let mode else { return nil }
        return current == mode ? .off : mode
    }

    /// The action that re-applies [mode]; `nil` for `.off` (Normal leaves the action unchanged).
    init?(mode: OneHandedMode) {
        switch mode {
        case .off: return nil
        case .left: self = .left
        case .right: self = .right
        }
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/core/settings/OneHandedMode.kt
package com.siansiansu.taigikeyboard.ime.core.settings

/**
 * One-handed keyboard mode: the key area narrows to [KEY_AREA_FRACTION] and
 * docks to one edge.
 *
 * Mirrors iOS `Settings/OneHandedMode.swift`; [storageValue] is the persisted
 * string under `PreferenceKeys.ONE_HANDED_MODE`, identical on both platforms.
 */
enum class OneHandedMode(
    val storageValue: String,
) {
    OFF("off"),
    LEFT("left"),
    RIGHT("right"),
    ;

    /** The opposite side; [OFF] stays [OFF]. */
    val flipped: OneHandedMode
        get() =
            when (this) {
                OFF -> OFF
                LEFT -> RIGHT
                RIGHT -> LEFT
            }

    companion object {
        // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/OneHandedMode.swift
        // keyAreaFraction (KeyboardKit's dock width). Drift causes silent divergence.
        const val KEY_AREA_FRACTION = 0.8f

        /** Coerce a stored raw string into a mode; unknown / absent values fall back to [OFF]. */
        fun fromStorage(raw: String?): OneHandedMode = entries.firstOrNull { it.storageValue == raw } ?: OFF
    }
}

/**
 * The action behind the toolbar keyboard button — the most recent pick from
 * its long-press callout (choosing Normal there leaves it unchanged).
 *
 * Mirrors iOS `KeyboardToolbarAction`; [storageValue] is the persisted string
 * under `PreferenceKeys.KEYBOARD_TOOLBAR_ACTION`.
 */
enum class KeyboardToolbarAction(
    val storageValue: String,
    /** The one-handed mode this action docks to; `null` for [DISMISS]. */
    val mode: OneHandedMode?,
) {
    DISMISS("dismiss", null),
    LEFT("left", OneHandedMode.LEFT),
    RIGHT("right", OneHandedMode.RIGHT),
    ;

    /** Mode after a tap on the toolbar button while in [current]; `null` = dismiss the keyboard.
     *  A side action toggles between that side and [OneHandedMode.OFF]. */
    fun tapResult(current: OneHandedMode): OneHandedMode? {
        val target = mode ?: return null
        return if (current == target) OneHandedMode.OFF else target
    }

    companion object {
        /** Coerce a stored raw string into an action; unknown / absent values fall back to [DISMISS]. */
        fun fromStorage(raw: String?): KeyboardToolbarAction = entries.firstOrNull { it.storageValue == raw } ?: DISMISS

        /** The action that re-applies [mode]; `null` for [OneHandedMode.OFF] (Normal leaves the action unchanged). */
        fun forMode(mode: OneHandedMode): KeyboardToolbarAction? = entries.firstOrNull { it.mode == mode }
    }
}
